soc: rp2350: enable support for building with CONFIG_XIP=n

dsseng opened the pull request:
Should be also possible on RP2040 but has less practical applications due to lack of UART boot.

Usage: `west build -b rpi_pico2/rp2350a/m33 samples/subsys/usb/console -- -DCONFIG_XIP=n`

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng commented:
Doesn't regress building without the snippet as nothing but commit hash changes in the built image

dsseng commented on snippets/rp2350-ram-boot/rp2350-ram-boot.overlay:
```diff
@@ -0,0 +1,59 @@
+
+#define RP2350_RAM_TOTAL_SIZE DT_SIZE_K(520)
+#define RP2350_RAM_CODE_SIZE  DT_SIZE_K(256)
+// Has to be adjusted as 0x20000000 + RP2350_RAM_CODE_SIZE if code partition is resized
+#define RP2350_RAM_OFFSET     20040000
```
Is there a less ugly solution to plug a computed value in `memory@RP2350_RAM_OFFSET` unit address?

dsseng replied:
Current code resize:
```diff
 #define RP2350_RAM_TOTAL_SIZE DT_SIZE_K(520)
-#define RP2350_RAM_CODE_SIZE  DT_SIZE_K(256)
+#define RP2350_RAM_CODE_SIZE  DT_SIZE_K(64)
 // Has to be adjusted as 0x20000000 + RP2350_RAM_CODE_SIZE if code partition is resized
-#define RP2350_RAM_OFFSET     20040000
+#define RP2350_RAM_OFFSET     20010000
 #define RP2350_IMAGE_DEF_SIZE 0x100
```

tejlmand commented:
Did you look at: https://github.com/zephyrproject-rtos/zephyr/discussions/36800

dsseng commented:
> Did you look at: #36800

Hm, I briefly read it now. Seems they also ended up patching DT, so this is basically it, but done as a snippet with an overlay to be applied easily 

Should I try just XIP=n to see if Zephyr automatically ignores flash and links into RAM? The only issue is that we have to add the image header to the RP2350 executable, thus we must use the ld script to add that section

dsseng commented:
> Did you look at: #36800

@tejlmand Could you please elaborate? I am doing basically the same. What should I give more attention to? Thanks

tejlmand reviewed (changes requested):
Having a ram boot / code location snippet could make sense., but it should be a generic snippet, where there could be a specific entry for handling of `rpi_pico2`, see https://docs.zephyrproject.org/latest/build/snippets/writing.html#board-specific-settings on how to do so in a snippet.

However, as commented, then it seems that ram code size is very depending on the sample / application size, and thus any size defined in the snippet would likely have to be adjusted.

For this reason, then perhaps it would be better to have a new test or sample which shows how a sample can be built and executed from flash / or placed directy in ram without using Zephyr's relocation feature.

Then users can build as:
```
$ west build -b rpi_pico2/rp2350a/m33 <new-sample>
```
or applying the overlay that could live together with the sample, like this:
```
$ west build -b rpi_pico2/rp2350a/m33 <new-sample> -- -DDTC_OVERLAY_FILE=ram.overlay
```

An example of such approach can be seen in https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/fs/format/ where a `ramdisk.overlay` overlay file is provided that the user can specify.
The advantage of dtc overlay approach is that it becomes more visible to users that the `ram.overlay` file should be copied to their own sample, and that they won't have to modify an upstream snippet.
The disadvantage of such approach compared to the snippet approach is of course that it's less visible that the ram option exists for any sample.

dsseng commented:
> Having a ram boot / code location snippet could make sense., but it should be a generic snippet, where there could be a specific entry for handling of `rpi_pico2`, see [docs.zephyrproject.org/latest/build/snippets/writing.html#board-specific-settings](https://docs.zephyrproject.org/latest/build/snippets/writing.html#board-specific-settings) on how to do so in a snippet.

Okay, perhaps with a regex I can match `*rp2350*` as this is applicable for all boards on this chip

> However, as commented, then it seems that ram code size is very depending on the sample / application size, and thus any size defined in the snippet would likely have to be adjusted

It is not likely to have to be adjusted. 256 KB "flash" and 256 KB RAM should be fine for most cases and others should be able to make a custom overlay/dts.

> The disadvantage of such approach compared to the snippet approach is of course that it's less visible that the ram option exists for any sample.

Yes, and this is more complex than a snippet for the development workflow.

I will probably implement a SoC-specific overlay in the snippet, so that we match not only `rpi_pico2/rp2350a/m33` but e.g. `adafruit_feather_rp2350/rp2350a/hazard3` or any other case. Resizing partitions should not be needed as the memory seems rather abundant for most use cases

dsseng committed:
soc: rp2350: add IMAGE_DEF using rom_start linker script

Use this approach for a simpler flow and to make relocation of images
into RAM easier.

Also do not force-select CONFIG_XIP (which is a default anyway), since
RP2350 can boot from SRAM.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
dts: boards: various: remove code-partition for RP2350

Updated linker scripts no longer necessitate that.

Users who employ a bootloader in their design should be able to adjust
partitioning more easily, removing IMAGE_DEF with a Kconfig if needed.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

tejlmand commented:
> Hm, I actually did some more work on XIP=n build for RP2, and perhaps I should just do a bit of a different thing. Please wait, this PR will be repurposed to make XIP=n build a bootable image.

perhaps marking the PR `DNM` or convert to draft meanwhile.

ajf58 reviewed (approved):
This is a neater solution than what I'd done originally (and was subsequently copied around). LGTM.

nordicjm reviewed (approved):
DTS changes OK

dsseng commented:
@tejlmand could you please take a look? These changes already got 3 approvals, so perhaps it'd be best to merge them to avoid unnecessary stale PRs and merge conflicts. Thank you!
